File: src/main/java/io/sc3/plethora/util/RangedInventoryWrapper.java
package pw.switchcraft.plethora.util;

import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.inventory.Inventory;
import net.minecraft.item.ItemStack;

public class RangedInventoryWrapper implements Inventory {
    private final Inventory underlying;
    private final int minSlot;
    private final int size;
    private final int maxSlot;

    public RangedInventoryWrapper(Inventory underlying, int minSlot, int size) {
        this.underlying = underlying;
        this.minSlot = minSlot;
        this.size = size;
        this.maxSlot = minSlot + size;
    }

    @Override
    public int size() {
        return size;
    }

    @Override
    public boolean isEmpty() {
        return false;
    }

    @Override
    public ItemStack getStack(int slot) {
        if (checkSlot(slot)) return underlying.getStack(slot + minSlot);
        return ItemStack.EMPTY;
    }

    @Override
    public ItemStack removeStack(int slot, int amount) {
        if (checkSlot(slot)) return underlying.removeStack(slot + minSlot, amount);
        return ItemStack.EMPTY;
    }

    @Override
    public ItemStack removeStack(int slot) {
        if (checkSlot(slot)) return underlying.removeStack(slot + minSlot);
        return ItemStack.EMPTY;
    }

    @Override
    public void setStack(int slot, ItemStack stack) {
        if (checkSlot(slot)) underlying.setStack(slot + minSlot, stack);
    }

    @Override
    public void markDirty() {
        underlying.markDirty();
    }

    @Override
    public boolean canPlayerUse(PlayerEntity player) {
        return underlying.canPlayerUse(player);
    }

    @Override
    public int getMaxCountPerStack() {
        return underlying.getMaxCountPerStack();
    }

    @Override
    public void clear() {
        throw new UnsupportedOperationException();
    }

    private boolean checkSlot(int localSlot) {
        return localSlot + minSlot < maxSlot;
    }
}

File: src/main/java/io/sc3/plethora/util/RangedInventoryWrapper.kt
package io.sc3.plethora.util

import io.sc3.plethora.api.reference.IReference
import net.minecraft.entity.LivingEntity
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.inventory.Inventory
import net.minecraft.item.ItemStack
import net.minecraft.item.ItemStack.EMPTY

class RangedInventoryWrapper<T : LivingEntity>(
  entityReference: IReference<T>,
  private val minSlot: Int,
  private val size: Int,
  private val inventoryProvider: (T) -> Inventory?
) : LivingEntityInventoryWrapper<T>(entityReference), Inventory {
  private val maxSlot = minSlot + size

  override fun size() = size
  override fun isEmpty() = false
  override fun getMaxCountPerStack() = safeInv?.maxCountPerStack ?: 64

  override fun canPlayerUse(player: PlayerEntity) =
    safeInv?.canPlayerUse(player) ?: false

  private fun checkSlot(localSlot: Int) =
    localSlot + minSlot < maxSlot

  override fun getStack(slot: Int): ItemStack =
    if (checkSlot(slot)) inv.getStack(slot + minSlot) else EMPTY

  override fun removeStack(slot: Int, amount: Int): ItemStack =
    if (checkSlot(slot)) inv.removeStack(slot + minSlot, amount) else EMPTY

  override fun removeStack(slot: Int): ItemStack =
    if (checkSlot(slot)) inv.removeStack(slot + minSlot) else EMPTY

  override fun setStack(slot: Int, stack: ItemStack) {
    if (checkSlot(slot)) inv.setStack(slot + minSlot, stack)
  }

  override fun markDirty() {
    inv.markDirty()
  }

  override fun clear() {
    throw UnsupportedOperationException()
  }

  override fun provideInventory(entity: T): Inventory? =
    inventoryProvider(entity)
}
